libraries/rush-lib/src/utilities/WebClient.ts

```diff
@@ -53,49 +56,124 @@ export enum WebClientProxy {
   Detect,
   Fiddler
 }
+export interface IRequestOptions extends RequestOptions, Pick<IFetchOptionsWithBody, 'body' | 'redirect'> {}
+
+export type FetchFn = (
+  url: string,
+  options: IRequestOptions,
+  isRedirect?: boolean
+) => Promise<IWebClientResponse>;
+
+const makeRequestAsync: FetchFn = async (
+  url: string,
+  options: IRequestOptions,
+  redirected: boolean = false
+) => {
+  const { body, redirect } = options;
+
+  return await new Promise(
+    (resolve: (result: IWebClientResponse) => void, reject: (error: Error) => void) => {
+      const parsedUrl: URL = typeof url === 'string' ? new URL(url) : url;
+      const requestFunction: typeof httpRequest | typeof httpsRequest =
+        parsedUrl.protocol === 'https:' ? httpsRequest : httpRequest;
+
+      requestFunction(url, options, (response: IncomingMessage) => {
+        const responseBuffers: (Buffer | Uint8Array)[] = [];
+        response.on('data', (chunk: string | Buffer | Uint8Array) => {
+          responseBuffers.push(Buffer.from(chunk));
```

Comment:
I feel like David would have comments on how we could do this in a more memory-efficient way, but I'm not sure of what that would be. I could also totally see someone wanting to pass in their own handler so that it can be written directly to file without touching memory. Though not sure if either of those need to be present in this PR since we know that this works.

Reply:
Will consider in a follow-up. @dmichon-msft - feedback is welcome.